improve error message when gradle tasks failing

### Steps to reproduce

This happened after this PR https://github.com/flutter/flutter/pull/153795 update the android sdk version of flutter.groovy to 35.

There should be tests to catch failures and prevent regression.


Error message:
```
You are applying Flutter's app_plugin_loader Gradle plugin imperatively using the apply script method, which is deprecated and will be removed in a future release. Migrate to applying Gradle plugins with the declarative plugins block: https://flutter.dev/to/flutter-gradle-plugin-apply

You are applying Flutter's main Gradle plugin imperatively using the apply script method, which is deprecated and will be removed in a future release. Migrate to applying Gradle plugins with the declarative plugins block: https://flutter.dev/to/flutter-gradle-plugin-apply


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:processReleaseResources'.
> A failure occurred while executing com.android.build.gradle.internal.res.LinkApplicationAndroidResourcesTask$TaskAction
   > Android resource linking failed
     aapt2 E 08-21 15:55:11 84751 7602282 LoadedArsc.cpp:94] RES_TABLE_TYPE_TYPE entry offsets overlap actual entry data.
     aapt2 E 08-21 15:55:11 84751 7602282 ApkAssets.cpp:149] Failed to load resources table in APK '/Users/jinhangyu/Library/Android/sdk/platforms/android-35/android.jar'.
     error: failed to load include path /Users/jinhangyu/Library/Android/sdk/platforms/android-35/android.jar.


* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 3s


You are applying Flutter's app_plugin_loader Gradle plugin imperatively using the apply script method, which is deprecated and will be
removed in a future release. Migrate to applying Gradle plugins with the declarative plugins block:
https://flutter.dev/to/flutter-gradle-plugin-apply

You are applying Flutter's main Gradle plugin imperatively using the apply script method, which is deprecated and will be removed in a future
release. Migrate to applying Gradle plugins with the declarative plugins block: https://flutter.dev/to/flutter-gradle-plugin-apply


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:processReleaseResources'.
> A failure occurred while executing com.android.build.gradle.internal.res.LinkApplicationAndroidResourcesTask$TaskAction
   > Android resource linking failed
     aapt2 E 08-21 15:55:11 84751 7602282 LoadedArsc.cpp:94] RES_TABLE_TYPE_TYPE entry offsets overlap actual entry data.
     aapt2 E 08-21 15:55:11 84751 7602282 ApkAssets.cpp:149] Failed to load resources table in APK
     '/Users/jinhangyu/Library/Android/sdk/platforms/android-35/android.jar'.
     error: failed to load include path /Users/jinhangyu/Library/Android/sdk/platforms/android-35/android.jar.


* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 3s

```

I ran into a different error, which can be solved by running a `flutter pub get`

<details>

```log
/Users/chtai/apps/clean/.dart_tool/package_config.json does not exist.
Did you run this command from the same directory as your pubspec.yaml file?
Target gen_dart_plugin_registrant failed: Error: 


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:compileFlutterBuildRelease'.
> Process 'command '/Users/chtai/git/flutter/bin/flutter'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 739ms


/Users/chtai/apps/clean/.dart_tool/package_config.json does not exist.
Did you run this command from the same directory as your pubspec.yaml file?
Target gen_dart_plugin_registrant failed: Error: 


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:compileFlutterBuildRelease'.
> Process 'command '/Users/chtai/git/flutter/bin/flutter'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 739ms
```

</details>

--- a/packages/flutter_tools/lib/src/android/gradle.dart
+++ b/packages/flutter_tools/lib/src/android/gradle.dart
@@ -259,6 +259,7 @@ class AndroidGradleBuilder implements AndroidBuilder {
     List<String> options = const <String>[],
     required FlutterProject project,
     required List<GradleHandledError> localGradleErrors,
+    required String gradleExecutablePath,
     int retry = 0,
     VoidCallback? preRunTask,
     VoidCallback? postRunTask,
@@ -309,7 +310,7 @@ class AndroidGradleBuilder implements AndroidBuilder {
       "Running Gradle task '$taskName'...",
     );
     final List<String> command = <String>[
-      _gradleUtils.getExecutable(project),
+      gradleExecutablePath,
       ...options, // suppresses gradle output.
       taskName,
     ];
@@ -361,6 +362,7 @@ class AndroidGradleBuilder implements AndroidBuilder {
               preRunTask: preRunTask,
               postRunTask: postRunTask,
               localGradleErrors: localGradleErrors,
+              gradleExecutablePath: gradleExecutablePath,
               retry: retry,
               project: project,
               maxRetries: maxRetries,
@@ -427,6 +429,8 @@ class AndroidGradleBuilder implements AndroidBuilder {
         project: project, buildInfo: androidBuildInfo.buildInfo);
     
     final List<String> options = <String>[];
+
+    final String gradleExecutablePath = _gradleUtils.getExecutable(project);
 
     // All automatically created files should exist.
     if (configOnly) {
@@ -536,7 +540,8 @@ class AndroidGradleBuilder implements AndroidBuilder {
         options: options,
         project: project,
         maxRetries: maxRetries,
-        localGradleErrors: localGradleErrors
+        localGradleErrors: localGradleErrors,
+        gradleExecutablePath: gradleExecutablePath,
     );
 
     if (exitCode != 0) {
@@ -809,6 +814,7 @@ class AndroidGradleBuilder implements AndroidBuilder {
           options: const <String>['-q'],
           project: project,
           localGradleErrors: gradleErrors,
+          gradleExecutablePath: _gradleUtils.getExecutable(project),
         outputParser: (String line) {
           if (_kBuildVariantRegex.firstMatch(line) case final RegExpMatch match) {
             results.add(match.namedGroup(_kBuildVariantRegexGroupName)!);
@@ -857,6 +863,7 @@ class AndroidGradleBuilder implements AndroidBuilder {
         options: <String>['-q', '-PoutputPath=$outputPath'],
         project: project,
         localGradleErrors: gradleErrors,
+        gradleExecutablePath: _gradleUtils.getExecutable(project),
       );
     } on Error catch(error) {
       _logger.printError(error.toString());

--- a/packages/flutter_tools/lib/src/android/gradle.dart
+++ b/packages/flutter_tools/lib/src/android/gradle.dart
@@ -427,7 +427,7 @@ class AndroidGradleBuilder implements AndroidBuilder {
     // from the local.properties file.
     updateLocalProperties(
         project: project, buildInfo: androidBuildInfo.buildInfo);
-    
+
     final List<String> options = <String>[];
 
     final String gradleExecutablePath = _gradleUtils.getExecutable(project);
@@ -821,7 +821,7 @@ class AndroidGradleBuilder implements AndroidBuilder {
           }
         }
       );
-    } on Error catch(error) {
+    } on Error catch (error) {
       _logger.printError(error.toString());
     }
 
@@ -865,7 +865,7 @@ class AndroidGradleBuilder implements AndroidBuilder {
         localGradleErrors: gradleErrors,
         gradleExecutablePath: _gradleUtils.getExecutable(project),
       );
-    } on Error catch(error) {
+    } on Error catch (error) {
       _logger.printError(error.toString());
     }
 

--- a/packages/flutter_tools/lib/src/android/gradle.dart
+++ b/packages/flutter_tools/lib/src/android/gradle.dart
@@ -53,6 +53,8 @@ final RegExp _kBuildVariantRegex = RegExp('^BuildVariant: (?<$_kBuildVariantRege
 const String _kBuildVariantRegexGroupName = 'variant';
 const String _kBuildVariantTaskName = 'printBuildVariants';
 
+typedef _OutputParser = void Function(String line);
+
 String _getOutputAppLinkSettingsTaskFor(String buildVariant) {
   return _taskForBuildVariant('output', buildVariant, 'AppLinkSettings');
 }
@@ -252,14 +254,15 @@ class AndroidGradleBuilder implements AndroidBuilder {
     );
   }
 
-  Future<RunResult> _runGradleTask(
-      String taskName, {
-        List<String> options = const <String>[],
-        required FlutterProject project,
-        required List<GradleHandledError> localGradleErrors,
-        int retry = 0,
-        int? maxRetries,
-      }) async {
+  Future<int> _runGradleTask(
+    String taskName, {
+    List<String> options = const <String>[],
+    required FlutterProject project,
+    required List<GradleHandledError> localGradleErrors,
+    int retry = 0,
+    int? maxRetries,
+    _OutputParser? outputParser,
+  }) async {
     final bool usesAndroidX = isAppUsingAndroidX(project.android.hostAppGradleRoot);
     if (usesAndroidX) {
       _analytics.send(Event.flutterBuildInfo(label: 'app-using-android-x', buildType: 'gradle'));
@@ -277,6 +280,9 @@ class AndroidGradleBuilder implements AndroidBuilder {
     GradleHandledError? detectedGradleError;
     String? detectedGradleErrorLine;
     String? consumeLog(String line) {
+      if (outputParser != null) {
+        outputParser(line);
+      }
       // The log lines that trigger incompatibleKotlinVersionHandler don't
       // always indicate an error, and there are times that that handler
       // covers up a more important error handler. Uniquely set it to be
@@ -306,22 +312,29 @@ class AndroidGradleBuilder implements AndroidBuilder {
       taskName,
     ];
 
-    RunResult result;
+    int exitCode = 1;
     try {
-      result = await _processUtils.run(
+      exitCode = await _processUtils.stream(
         command,
         workingDirectory: project.android.hostAppGradleRoot.path,
         allowReentrantFlutter: true,
         environment: _java?.environment,
+        mapFunction: consumeLog,
       );
+    } on ProcessException catch (exception) {
+      consumeLog(exception.toString());
+      // Rethrow the exception if the error isn't handled by any of the
+      // `localGradleErrors`.
+      if (detectedGradleError == null) {
+        rethrow;
+      }
     } finally {
       status.stop();
     }
-    if (result.exitCode != 0) {
-      consumeLog(result.stderr);
+    if (exitCode != 0) {
       if (detectedGradleError == null) {
         _analytics.send(Event.flutterBuildInfo(label: 'gradle-unknown-failure', buildType: 'gradle'));
-        return result;
+        return exitCode;
       }
       final GradleBuildStatus status = await detectedGradleError!.handler(
         line: detectedGradleErrorLine!,
@@ -338,7 +351,7 @@ class AndroidGradleBuilder implements AndroidBuilder {
             retry += 1;
             _logger.printStatus('Retrying Gradle Build: #$retry, wait time: ${waitTime}ms');
             await Future<void>.delayed(Duration(milliseconds: waitTime));
-            result = await _runGradleTask(
+            exitCode = await _runGradleTask(
               taskName,
               options: options,
               localGradleErrors: localGradleErrors,
@@ -346,11 +359,11 @@ class AndroidGradleBuilder implements AndroidBuilder {
               project: project,
               maxRetries: maxRetries,
             );
-            if (result.exitCode == 0) {
+            if (exitCode == 0) {
               final String successEventLabel = 'gradle-${detectedGradleError!.eventLabel}-success';
 
               _analytics.send(Event.flutterBuildInfo(label: successEventLabel, buildType: 'gradle'));
-              return result;
+              return exitCode;
             }
           case GradleBuildStatus.exit:
           // Continue and throw tool exit.
@@ -359,7 +372,7 @@ class AndroidGradleBuilder implements AndroidBuilder {
       final String usageLabel = 'gradle-${detectedGradleError?.eventLabel}-failure';
       _analytics.send(Event.flutterBuildInfo(label: usageLabel, buildType: 'gradle'));
     }
-    return result;
+    return exitCode;
   }
 
   /// Builds an app.
@@ -514,7 +527,7 @@ class AndroidGradleBuilder implements AndroidBuilder {
 
     final Stopwatch sw = Stopwatch()
       ..start();
-    final RunResult result = await _runGradleTask(
+    final int exitCode = await _runGradleTask(
         assembleTask,
         options: options,
         project: project,
@@ -528,7 +541,7 @@ class AndroidGradleBuilder implements AndroidBuilder {
       elapsedMilliseconds: elapsedDuration.inMilliseconds,
     ));
 
-    if (result.exitCode != 0) {
+    if (exitCode != 0) {
       throwToolExit(
         'Gradle task $assembleTask failed with exit code $exitCode',
         exitCode: exitCode,
@@ -779,12 +792,24 @@ class AndroidGradleBuilder implements AndroidBuilder {
   Future<List<String>> getBuildVariants({required FlutterProject project}) async {
     final Stopwatch sw = Stopwatch()
       ..start();
-    final RunResult result = await _runGradleTask(
-      _kBuildVariantTaskName,
-      options: const <String>['-q'],
-      project: project,
-      localGradleErrors: gradleErrors
-    );
+    int exitCode = 1;
+    final List<String> results = <String>[];
+
+    try {
+      exitCode = await _runGradleTask(
+          _kBuildVariantTaskName,
+          options: const <String>['-q'],
+          project: project,
+          localGradleErrors: gradleErrors,
+        outputParser: (String line) {
+          if (_kBuildVariantRegex.firstMatch(line) case final RegExpMatch match) {
+            results.add(match.namedGroup(_kBuildVariantRegexGroupName)!);
+          }
+        }
+      );
+    } on Error catch(error) {
+      _logger.printError(error.toString());
+    }
 
     final Duration elapsedDuration = sw.elapsed;
     _analytics.send(Event.timing(
@@ -793,15 +818,11 @@ class AndroidGradleBuilder implements AndroidBuilder {
       elapsedMilliseconds: elapsedDuration.inMilliseconds,
     ));
 
-    if (result.exitCode != 0) {
+    if (exitCode != 0) {
       return const <String>[];
     }
 
-    return <String>[
-      for (final String line in LineSplitter.split(result.stdout))
-        if (_kBuildVariantRegex.firstMatch(line) case final RegExpMatch match)
-          match.namedGroup(_kBuildVariantRegexGroupName)!,
-    ];
+    return results;
   }
 
   @override
@@ -818,12 +839,17 @@ class AndroidGradleBuilder implements AndroidBuilder {
     );
     final Stopwatch sw = Stopwatch()
       ..start();
-    final RunResult result = await _runGradleTask(
-      taskName,
-      options: <String>['-q', '-PoutputPath=$outputPath'],
-      project: project,
-      localGradleErrors: gradleErrors,
-    );
+    int exitCode = 1;
+    try {
+      exitCode = await _runGradleTask(
+        taskName,
+        options: <String>['-q', '-PoutputPath=$outputPath'],
+        project: project,
+        localGradleErrors: gradleErrors,
+      );
+    } on Error catch(error) {
+      _logger.printError(error.toString());
+    }
     final Duration elapsedDuration = sw.elapsed;
     _analytics.send(Event.timing(
       workflow: 'outputs',
@@ -831,8 +857,8 @@ class AndroidGradleBuilder implements AndroidBuilder {
       elapsedMilliseconds: elapsedDuration.inMilliseconds,
     ));
 
-    if (result.exitCode != 0) {
-      throwToolExit(result.stderr);
+    if (exitCode != 0) {
+      throwToolExit('Gradle task $taskName failed with exit code $exitCode');
     }
     return outputPath;
   }

--- a/packages/flutter_tools/lib/src/android/gradle.dart
+++ b/packages/flutter_tools/lib/src/android/gradle.dart
@@ -253,10 +253,50 @@ class AndroidGradleBuilder implements AndroidBuilder {
   }
 
   Future<RunResult> _runGradleTask(
-    String taskName, {
-    List<String> options = const <String>[],
-    required FlutterProject project
-  }) async {
+      String taskName, {
+        List<String> options = const <String>[],
+        required FlutterProject project,
+        required List<GradleHandledError> localGradleErrors,
+        int retry = 0,
+        int? maxRetries,
+      }) async {
+    final bool usesAndroidX = isAppUsingAndroidX(project.android.hostAppGradleRoot);
+    if (usesAndroidX) {
+      _analytics.send(Event.flutterBuildInfo(label: 'app-using-android-x', buildType: 'gradle'));
+    } else if (!usesAndroidX) {
+      _analytics.send(Event.flutterBuildInfo(label: 'app-not-using-android-x', buildType: 'gradle'));
+
+      _logger.printStatus("${_logger.terminal.warningMark} Your app isn't using AndroidX.", emphasis: true);
+      _logger.printStatus(
+        'To avoid potential build failures, you can quickly migrate your app '
+            'by following the steps on https://goo.gl/CP92wY .',
+        indent: 4,
+      );
+    }
+
+    GradleHandledError? detectedGradleError;
+    String? detectedGradleErrorLine;
+    String? consumeLog(String line) {
+      // The log lines that trigger incompatibleKotlinVersionHandler don't
+      // always indicate an error, and there are times that that handler
+      // covers up a more important error handler. Uniquely set it to be
+      // the lowest priority handler by allowing it to be overridden.
+      if (detectedGradleError != null
+          && detectedGradleError != incompatibleKotlinVersionHandler) {
+        // Pipe stdout/stderr from Gradle.
+        return line;
+      }
+      for (final GradleHandledError gradleError in localGradleErrors) {
+        if (gradleError.test(line)) {
+          detectedGradleErrorLine = line;
+          detectedGradleError = gradleError;
+          // The first error match wins.
+          break;
+        }
+      }
+      // Pipe stdout/stderr from Gradle.
+      return line;
+    }
     final Status status = _logger.startProgress(
       "Running Gradle task '$taskName'...",
     );
@@ -276,6 +316,48 @@ class AndroidGradleBuilder implements AndroidBuilder {
       );
     } finally {
       status.stop();
+    }
+    if (result.exitCode != 0) {
+      consumeLog(result.stderr);
+      if (detectedGradleError == null) {
+        _analytics.send(Event.flutterBuildInfo(label: 'gradle-unknown-failure', buildType: 'gradle'));
+        return result;
+      }
+      final GradleBuildStatus status = await detectedGradleError!.handler(
+        line: detectedGradleErrorLine!,
+        project: project,
+        usesAndroidX: usesAndroidX,
+      );
+
+      if (maxRetries == null || retry < maxRetries) {
+        switch (status) {
+          case GradleBuildStatus.retry:
+          // Use binary exponential backoff before retriggering the build.
+          // The expected wait times are: 100ms, 200ms, 400ms, and so on...
+            final int waitTime = min(pow(2, retry).toInt() * 100, kMaxRetryTime.inMicroseconds);
+            retry += 1;
+            _logger.printStatus('Retrying Gradle Build: #$retry, wait time: ${waitTime}ms');
+            await Future<void>.delayed(Duration(milliseconds: waitTime));
+            result = await _runGradleTask(
+              taskName,
+              options: options,
+              localGradleErrors: localGradleErrors,
+              retry: retry,
+              project: project,
+              maxRetries: maxRetries,
+            );
+            if (result.exitCode == 0) {
+              final String successEventLabel = 'gradle-${detectedGradleError!.eventLabel}-success';
+
+              _analytics.send(Event.flutterBuildInfo(label: successEventLabel, buildType: 'gradle'));
+              return result;
+            }
+          case GradleBuildStatus.exit:
+          // Continue and throw tool exit.
+        }
+      }
+      final String usageLabel = 'gradle-${detectedGradleError?.eventLabel}-failure';
+      _analytics.send(Event.flutterBuildInfo(label: usageLabel, buildType: 'gradle'));
     }
     return result;
   }
@@ -335,12 +417,8 @@ class AndroidGradleBuilder implements AndroidBuilder {
     // from the local.properties file.
     updateLocalProperties(
         project: project, buildInfo: androidBuildInfo.buildInfo);
-
-    final List<String> command = <String>[
-      // This does more than get gradlewrapper. It creates the file, ensures it
-      // exists and verifies the file is executable.
-      _gradleUtils.getExecutable(project),
-    ];
+    
+    final List<String> options = <String>[];
 
     // All automatically created files should exist.
     if (configOnly) {
@@ -354,22 +432,18 @@ class AndroidGradleBuilder implements AndroidBuilder {
         ? getBundleTaskFor(buildInfo)
         : getAssembleTaskFor(buildInfo);
 
-    final Status status = _logger.startProgress(
-      "Running Gradle task '$assembleTask'...",
-    );
-
     if (_logger.isVerbose) {
-      command.add('--full-stacktrace');
-      command.add('--info');
-      command.add('-Pverbose=true');
+      options.add('--full-stacktrace');
+      options.add('--info');
+      options.add('-Pverbose=true');
     } else {
-      command.add('-q');
+      options.add('-q');
     }
     if (!buildInfo.androidGradleDaemon) {
-      command.add('--no-daemon');
+      options.add('--no-daemon');
     }
     if (buildInfo.androidSkipBuildDependencyValidation) {
-      command.add('-PskipDependencyChecks=true');
+      options.add('-PskipDependencyChecks=true');
     }
     final LocalEngineInfo? localEngineInfo = _artifacts.localEngineInfo;
     if (localEngineInfo != null) {
@@ -382,29 +456,29 @@ class AndroidGradleBuilder implements AndroidBuilder {
           'Using local engine: ${localEngineInfo.targetOutPath}\n'
               'Local Maven repo: ${localEngineRepo.path}'
       );
-      command.add('-Plocal-engine-repo=${localEngineRepo.path}');
-      command.add('-Plocal-engine-build-mode=${buildInfo.modeName}');
-      command.add('-Plocal-engine-out=${localEngineInfo.targetOutPath}');
-      command.add('-Plocal-engine-host-out=${localEngineInfo.hostOutPath}');
-      command.add('-Ptarget-platform=${_getTargetPlatformByLocalEnginePath(
+      options.add('-Plocal-engine-repo=${localEngineRepo.path}');
+      options.add('-Plocal-engine-build-mode=${buildInfo.modeName}');
+      options.add('-Plocal-engine-out=${localEngineInfo.targetOutPath}');
+      options.add('-Plocal-engine-host-out=${localEngineInfo.hostOutPath}');
+      options.add('-Ptarget-platform=${_getTargetPlatformByLocalEnginePath(
           localEngineInfo.targetOutPath)}');
     } else if (androidBuildInfo.targetArchs.isNotEmpty) {
       final String targetPlatforms = androidBuildInfo
           .targetArchs
           .map((AndroidArch e) => e.platformName).join(',');
-      command.add('-Ptarget-platform=$targetPlatforms');
+      options.add('-Ptarget-platform=$targetPlatforms');
     }
-    command.add('-Ptarget=$target');
+    options.add('-Ptarget=$target');
     // If using v1 embedding, we want to use FlutterApplication as the base app.
     final String baseApplicationName =
         project.android.getEmbeddingVersion() == AndroidEmbeddingVersion.v2 ?
           'android.app.Application' :
           'io.flutter.app.FlutterApplication';
-    command.add('-Pbase-application-name=$baseApplicationName');
+    options.add('-Pbase-application-name=$baseApplicationName');
     final List<DeferredComponent>? deferredComponents = project.manifest.deferredComponents;
     if (deferredComponents != null) {
       if (deferredComponentsEnabled) {
-        command.add('-Pdeferred-components=true');
+        options.add('-Pdeferred-components=true');
         androidBuildInfo.buildInfo.dartDefines.add('validate-deferred-components=$validateDeferredComponents');
       }
       // Pass in deferred components regardless of building split aot to satisfy
@@ -414,76 +488,38 @@ class AndroidGradleBuilder implements AndroidBuilder {
         componentNames.add(component.name);
       }
       if (componentNames.isNotEmpty) {
-        command.add('-Pdeferred-component-names=${componentNames.join(',')}');
+        options.add('-Pdeferred-component-names=${componentNames.join(',')}');
         // Multi-apk applications cannot use shrinking. This is only relevant when using
         // android dynamic feature modules.
         _logger.printStatus(
           'Shrinking has been disabled for this build due to deferred components. Shrinking is '
           'not available for multi-apk applications. This limitation is expected to be removed '
           'when Gradle plugin 4.2+ is available in Flutter.', color: TerminalColor.yellow);
-        command.add('-Pshrink=false');
+        options.add('-Pshrink=false');
       }
     }
-    command.addAll(androidBuildInfo.buildInfo.toGradleConfig());
+    options.addAll(androidBuildInfo.buildInfo.toGradleConfig());
     if (buildInfo.fileSystemRoots.isNotEmpty) {
-      command.add('-Pfilesystem-roots=${buildInfo.fileSystemRoots.join('|')}');
+      options.add('-Pfilesystem-roots=${buildInfo.fileSystemRoots.join('|')}');
     }
     if (buildInfo.fileSystemScheme != null) {
-      command.add('-Pfilesystem-scheme=${buildInfo.fileSystemScheme}');
+      options.add('-Pfilesystem-scheme=${buildInfo.fileSystemScheme}');
     }
     if (androidBuildInfo.splitPerAbi) {
-      command.add('-Psplit-per-abi=true');
+      options.add('-Psplit-per-abi=true');
     }
     if (androidBuildInfo.fastStart) {
-      command.add('-Pfast-start=true');
-    }
-    command.add(assembleTask);
-
-    GradleHandledError? detectedGradleError;
-    String? detectedGradleErrorLine;
-    String? consumeLog(String line) {
-      // The log lines that trigger incompatibleKotlinVersionHandler don't
-      // always indicate an error, and there are times that that handler
-      // covers up a more important error handler. Uniquely set it to be
-      // the lowest priority handler by allowing it to be overridden.
-      if (detectedGradleError != null
-          && detectedGradleError != incompatibleKotlinVersionHandler) {
-        // Pipe stdout/stderr from Gradle.
-        return line;
-      }
-      for (final GradleHandledError gradleError in localGradleErrors) {
-        if (gradleError.test(line)) {
-          detectedGradleErrorLine = line;
-          detectedGradleError = gradleError;
-          // The first error match wins.
-          break;
-        }
-      }
-      // Pipe stdout/stderr from Gradle.
-      return line;
+      options.add('-Pfast-start=true');
     }
 
     final Stopwatch sw = Stopwatch()
       ..start();
-    int exitCode = 1;
-    try {
-      exitCode = await _processUtils.stream(
-        command,
-        workingDirectory: project.android.hostAppGradleRoot.path,
-        allowReentrantFlutter: true,
-        environment: _java?.environment,
-        mapFunction: consumeLog,
-      );
-    } on ProcessException catch (exception) {
-      consumeLog(exception.toString());
-      // Rethrow the exception if the error isn't handled by any of the
-      // `localGradleErrors`.
-      if (detectedGradleError == null) {
-        rethrow;
-      }
-    } finally {
-      status.stop();
-    }
+    final RunResult result = await _runGradleTask(
+        assembleTask,
+        options: options,
+        project: project,
+        localGradleErrors: gradleErrors
+    );
 
     final Duration elapsedDuration = sw.elapsed;
     _analytics.send(Event.timing(
@@ -492,52 +528,7 @@ class AndroidGradleBuilder implements AndroidBuilder {
       elapsedMilliseconds: elapsedDuration.inMilliseconds,
     ));
 
-    if (exitCode != 0) {
-      if (detectedGradleError == null) {
-        _analytics.send(Event.flutterBuildInfo(label: 'gradle-unknown-failure', buildType: 'gradle'));
-
-        throwToolExit(
-          'Gradle task $assembleTask failed with exit code $exitCode',
-          exitCode: exitCode,
-        );
-      }
-      final GradleBuildStatus status = await detectedGradleError!.handler(
-        line: detectedGradleErrorLine!,
-        project: project,
-        usesAndroidX: usesAndroidX,
-      );
-
-      if (maxRetries == null || retry < maxRetries) {
-        switch (status) {
-          case GradleBuildStatus.retry:
-            // Use binary exponential backoff before retriggering the build.
-            // The expected wait times are: 100ms, 200ms, 400ms, and so on...
-            final int waitTime = min(pow(2, retry).toInt() * 100, kMaxRetryTime.inMicroseconds);
-            retry += 1;
-            _logger.printStatus('Retrying Gradle Build: #$retry, wait time: ${waitTime}ms');
-            await Future<void>.delayed(Duration(milliseconds: waitTime));
-            await buildGradleApp(
-              project: project,
-              androidBuildInfo: androidBuildInfo,
-              target: target,
-              isBuildingBundle: isBuildingBundle,
-              localGradleErrors: localGradleErrors,
-              retry: retry,
-              maxRetries: maxRetries,
-              configOnly: configOnly,
-            );
-            final String successEventLabel = 'gradle-${detectedGradleError!.eventLabel}-success';
-
-            _analytics.send(Event.flutterBuildInfo(label: successEventLabel, buildType: 'gradle'));
-
-            return;
-          case GradleBuildStatus.exit:
-            // Continue and throw tool exit.
-        }
-      }
-      final String usageLabel = 'gradle-${detectedGradleError?.eventLabel}-failure';
-      _analytics.send(Event.flutterBuildInfo(label: usageLabel, buildType: 'gradle'));
-
+    if (result.exitCode != 0) {
       throwToolExit(
         'Gradle task $assembleTask failed with exit code $exitCode',
         exitCode: exitCode,
@@ -792,6 +783,7 @@ class AndroidGradleBuilder implements AndroidBuilder {
       _kBuildVariantTaskName,
       options: const <String>['-q'],
       project: project,
+      localGradleErrors: gradleErrors
     );
 
     final Duration elapsedDuration = sw.elapsed;
@@ -802,10 +794,9 @@ class AndroidGradleBuilder implements AndroidBuilder {
     ));
 
     if (result.exitCode != 0) {
-      _logger.printStatus(result.stdout, wrap: false);
-      _logger.printError(result.stderr, wrap: false);
       return const <String>[];
     }
+
     return <String>[
       for (final String line in LineSplitter.split(result.stdout))
         if (_kBuildVariantRegex.firstMatch(line) case final RegExpMatch match)
@@ -831,6 +822,7 @@ class AndroidGradleBuilder implements AndroidBuilder {
       taskName,
       options: <String>['-q', '-PoutputPath=$outputPath'],
       project: project,
+      localGradleErrors: gradleErrors,
     );
     final Duration elapsedDuration = sw.elapsed;
     _analytics.send(Event.timing(
@@ -840,8 +832,6 @@ class AndroidGradleBuilder implements AndroidBuilder {
     ));
 
     if (result.exitCode != 0) {
-      _logger.printStatus(result.stdout, wrap: false);
-      _logger.printError(result.stderr, wrap: false);
       throwToolExit(result.stderr);
     }
     return outputPath;

--- a/packages/flutter_tools/lib/src/android/gradle.dart
+++ b/packages/flutter_tools/lib/src/android/gradle.dart
@@ -260,6 +260,8 @@ class AndroidGradleBuilder implements AndroidBuilder {
     required FlutterProject project,
     required List<GradleHandledError> localGradleErrors,
     int retry = 0,
+    VoidCallback? preRunTask,
+    VoidCallback? postRunTask,
     int? maxRetries,
     _OutputParser? outputParser,
   }) async {
@@ -311,6 +313,7 @@ class AndroidGradleBuilder implements AndroidBuilder {
       ...options, // suppresses gradle output.
       taskName,
     ];
+    preRunTask?.call();
 
     int exitCode = 1;
     try {
@@ -331,6 +334,7 @@ class AndroidGradleBuilder implements AndroidBuilder {
     } finally {
       status.stop();
     }
+    postRunTask?.call();
     if (exitCode != 0) {
       if (detectedGradleError == null) {
         _analytics.send(Event.flutterBuildInfo(label: 'gradle-unknown-failure', buildType: 'gradle'));
@@ -354,24 +358,28 @@ class AndroidGradleBuilder implements AndroidBuilder {
             exitCode = await _runGradleTask(
               taskName,
               options: options,
+              preRunTask: preRunTask,
+              postRunTask: postRunTask,
               localGradleErrors: localGradleErrors,
               retry: retry,
               project: project,
               maxRetries: maxRetries,
             );
             if (exitCode == 0) {
-              final String successEventLabel = 'gradle-${detectedGradleError!.eventLabel}-success';
-
-              _analytics.send(Event.flutterBuildInfo(label: successEventLabel, buildType: 'gradle'));
+              final String successEventLabel = 'gradle-${detectedGradleError!
+                  .eventLabel}-success';
+              _analytics.send(Event.flutterBuildInfo(
+                  label: successEventLabel, buildType: 'gradle'));
               return exitCode;
             }
           case GradleBuildStatus.exit:
           // Continue and throw tool exit.
         }
       }
-      final String usageLabel = 'gradle-${detectedGradleError?.eventLabel}-failure';
+      final String usageLabel = 'gradle-${detectedGradleError!.eventLabel}-failure';
       _analytics.send(Event.flutterBuildInfo(label: usageLabel, buildType: 'gradle'));
     }
+
     return exitCode;
   }
 
@@ -413,19 +421,6 @@ class AndroidGradleBuilder implements AndroidBuilder {
     final ProjectMigration migration = ProjectMigration(migrators);
     await migration.run();
 
-    final bool usesAndroidX = isAppUsingAndroidX(project.android.hostAppGradleRoot);
-    if (usesAndroidX) {
-      _analytics.send(Event.flutterBuildInfo(label: 'app-using-android-x', buildType: 'gradle'));
-    } else if (!usesAndroidX) {
-      _analytics.send(Event.flutterBuildInfo(label: 'app-not-using-android-x', buildType: 'gradle'));
-
-      _logger.printStatus("${_logger.terminal.warningMark} Your app isn't using AndroidX.", emphasis: true);
-      _logger.printStatus(
-        'To avoid potential build failures, you can quickly migrate your app '
-            'by following the steps on https://goo.gl/CP92wY .',
-        indent: 4,
-      );
-    }
     // The default Gradle script reads the version name and number
     // from the local.properties file.
     updateLocalProperties(
@@ -524,22 +519,25 @@ class AndroidGradleBuilder implements AndroidBuilder {
     if (androidBuildInfo.fastStart) {
       options.add('-Pfast-start=true');
     }
-
-    final Stopwatch sw = Stopwatch()
-      ..start();
+    late Stopwatch sw;
     final int exitCode = await _runGradleTask(
         assembleTask,
+        preRunTask: () {
+          sw = Stopwatch()..start();
+        },
+        postRunTask: () {
+          final Duration elapsedDuration = sw.elapsed;
+          _analytics.send(Event.timing(
+            workflow: 'build',
+            variableName: 'gradle',
+            elapsedMilliseconds: elapsedDuration.inMilliseconds,
+          ));
+        },
         options: options,
         project: project,
-        localGradleErrors: gradleErrors
+        maxRetries: maxRetries,
+        localGradleErrors: localGradleErrors
     );
-
-    final Duration elapsedDuration = sw.elapsed;
-    _analytics.send(Event.timing(
-      workflow: 'build',
-      variableName: 'gradle',
-      elapsedMilliseconds: elapsedDuration.inMilliseconds,
-    ));
 
     if (exitCode != 0) {
       throwToolExit(
@@ -790,14 +788,24 @@ class AndroidGradleBuilder implements AndroidBuilder {
 
   @override
   Future<List<String>> getBuildVariants({required FlutterProject project}) async {
-    final Stopwatch sw = Stopwatch()
-      ..start();
+    late Stopwatch sw;
     int exitCode = 1;
     final List<String> results = <String>[];
 
     try {
       exitCode = await _runGradleTask(
           _kBuildVariantTaskName,
+          preRunTask: () {
+            sw = Stopwatch()..start();
+          },
+          postRunTask: () {
+            final Duration elapsedDuration = sw.elapsed;
+            _analytics.send(Event.timing(
+              workflow: 'print',
+              variableName: 'android build variants',
+              elapsedMilliseconds: elapsedDuration.inMilliseconds,
+            ));
+          },
           options: const <String>['-q'],
           project: project,
           localGradleErrors: gradleErrors,
@@ -810,13 +818,6 @@ class AndroidGradleBuilder implements AndroidBuilder {
     } on Error catch(error) {
       _logger.printError(error.toString());
     }
-
-    final Duration elapsedDuration = sw.elapsed;
-    _analytics.send(Event.timing(
-      workflow: 'print',
-      variableName: 'android build variants',
-      elapsedMilliseconds: elapsedDuration.inMilliseconds,
-    ));
 
     if (exitCode != 0) {
       return const <String>[];
@@ -837,12 +838,22 @@ class AndroidGradleBuilder implements AndroidBuilder {
       directory.absolute.path,
       'app-link-settings-$buildVariant.json',
     );
-    final Stopwatch sw = Stopwatch()
-      ..start();
+    late Stopwatch sw;
     int exitCode = 1;
     try {
       exitCode = await _runGradleTask(
         taskName,
+        preRunTask: () {
+          sw = Stopwatch()..start();
+        },
+        postRunTask: () {
+          final Duration elapsedDuration = sw.elapsed;
+          _analytics.send(Event.timing(
+            workflow: 'outputs',
+            variableName: 'app link settings',
+            elapsedMilliseconds: elapsedDuration.inMilliseconds,
+          ));
+        },
         options: <String>['-q', '-PoutputPath=$outputPath'],
         project: project,
         localGradleErrors: gradleErrors,
@@ -850,12 +861,6 @@ class AndroidGradleBuilder implements AndroidBuilder {
     } on Error catch(error) {
       _logger.printError(error.toString());
     }
-    final Duration elapsedDuration = sw.elapsed;
-    _analytics.send(Event.timing(
-      workflow: 'outputs',
-      variableName: 'app link settings',
-      elapsedMilliseconds: elapsedDuration.inMilliseconds,
-    ));
 
     if (exitCode != 0) {
       throwToolExit('Gradle task $taskName failed with exit code $exitCode');

--- a/packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart
+++ b/packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart
@@ -339,8 +339,7 @@ void main() {
       expect(logger.statusText, contains('Retrying Gradle Build: #2, wait time: 200ms'));
 
       expect(testFnCalled, equals(maxRetries + 1));
-
-      expect(fakeAnalytics.sentEvents, hasLength(7));
+      expect(fakeAnalytics.sentEvents, hasLength(9));
       expect(fakeAnalytics.sentEvents, contains(
         Event.flutterBuildInfo(
           label: 'gradle-random-event-label-failure',

